File: src/agent_dump/agents/claude_transcript.py
from dataclasses import dataclass, field
from typing import Any

from agent_dump.agents.jsonl_scan import parse_iso_timestamp_ms
from agent_dump.agents.message_assembly import (
    backfill_tool_state,
    build_fallback_tool_message,
    build_message,
    build_text_part,
    build_tool_part,
    try_append_to_assistant_group,
)
from agent_dump.agents.message_types import NormalizedMessage, NormalizedPart
from agent_dump.coercion import safe_int


@dataclass
class _ClaudeTranscriptState:
    messages: list[NormalizedMessage] = field(default_factory=list)
    pending_tool_calls: dict[str, tuple[int, int]] = field(default_factory=dict)
    ignored_tool_call_ids: set[str] = field(default_factory=set)
    assistant_uuid_to_tool_calls: dict[str, list[str]] = field(default_factory=dict)
    current_assistant_index: int | None = None
    latest_assistant_text_index: int | None = None

    def reset_assistant_group(self) -> None:
        self.current_assistant_index = None
        self.latest_assistant_text_index = None


class ClaudeTranscriptDecoder:
    """Decode Claude JSONL records into normalized transcript messages."""

    def __init__(self) -> None:
        self._state = _ClaudeTranscriptState()

    @property
    def messages(self) -> list[NormalizedMessage]:
        return self._state.messages

    def append_record(self, data: dict[str, Any]) -> None:
        if data.get("isMeta") is True:
            return

        msg_type = data.get("type", "")
        if msg_type == "assistant":
            self._append_assistant_record(data)
            return
        if msg_type == "user":
            self._append_user_record(data)
            return
        if msg_type != "tool_result":
            return

        timestamp_ms = self._parse_timestamp_ms(data)
        msg = data.get("message", {})
        output_parts = self._normalize_tool_output(msg.get("content"), timestamp_ms)
        fallback_message = build_fallback_tool_message(
            message_id=str(data.get("uuid", "")),
            output_parts=output_parts,
            time_created=timestamp_ms,
        )
        if fallback_message:
            self._state.messages.append(fallback_message)
        self._state.reset_assistant_group()

    def token_totals(self) -> tuple[int, int]:
        input_tokens = 0
        output_tokens = 0
        for message in self._state.messages:
            tokens = message.get("tokens")
            if not isinstance(tokens, dict):
                continue
            input_tokens += safe_int(tokens.get("input_tokens"))
            output_tokens += safe_int(tokens.get("output_tokens"))
        return input_tokens, output_tokens

    def _parse_timestamp_ms(self, data: dict[str, Any]) -> int:
        return parse_iso_timestamp_ms(data.get("timestamp"))

    def _build_tool_part(self, part: dict[str, Any], timestamp_ms: int) -> NormalizedPart:
        tool_name = str(part.get("name", ""))
        return build_tool_part(
            tool_name=tool_name,
            call_id=str(part.get("id", "")),
            title=f"Tool: {tool_name}",
            state={"input": part.get("input", {}), "output": None},
            timestamp_ms=timestamp_ms,
        )

    def _normalize_tool_output(self, content: Any, timestamp_ms: int) -> list[NormalizedPart]:
        if isinstance(content, str):
            return [build_text_part(content, timestamp_ms)] if content.strip() else []
        if isinstance(content, list):
            parts: list[NormalizedPart] = []
            for item in content:
                if isinstance(item, dict):
                    text = str(item.get("text", item.get("content", "")))
                    if text.strip():
                        parts.append(build_text_part(text, timestamp_ms))
                elif isinstance(item, str) and item.strip():
                    parts.append(build_text_part(item, timestamp_ms))
            return parts
        if content is None:
            return []
        text = str(content)
        return [build_text_part(text, timestamp_ms)] if text.strip() else []

    def _normalize_user_text_parts(self, content: Any, timestamp_ms: int) -> list[NormalizedPart]:
        if isinstance(content, str):
            return [build_text_part(content, timestamp_ms)] if content.strip() else []
        if not isinstance(content, list):
            return []

        parts: list[NormalizedPart] = []
        for item in content:
            if isinstance(item, dict):
                if item.get("type") == "tool_result":
                    continue
                text = str(item.get("text", ""))
                if text.strip():
                    parts.append(build_text_part(text, timestamp_ms))
            elif isinstance(item, str) and item.strip():
                parts.append(build_text_part(item, timestamp_ms))
        return parts

    def _apply_assistant_metadata(self, message: NormalizedMessage, msg: dict[str, Any]) -> None:
        model = msg.get("model")
        usage = msg.get("usage")
        if model and not message.get("model"):
            message["model"] = model
        if isinstance(usage, dict) and not message.get("tokens"):
            message["tokens"] = usage

    def _append_assistant_reasoning(
        self,
        *,
        message_id: str,
        msg: dict[str, Any],
        timestamp_ms: int,
        part: NormalizedPart,
    ) -> int:
        state = self._state
        folded = try_append_to_assistant_group(
            state.messages,
            current_assistant_index=state.current_assistant_index,
            parts=(part,),
            blocking_part_types=("text", "tool"),
            on_message=lambda existing: self._apply_assistant_metadata(existing, msg),
        )
        if folded is not None:
            return folded

        message = build_message(
            message_id=message_id,
            role="assistant",
            agent="claude",
            time_created=timestamp_ms,
            parts=[part],
        )
        self._apply_assistant_metadata(message, msg)
        state.messages.append(message)
        return len(state.messages) - 1

    def _append_assistant_text(
        self,
        *,
        message_id: str,
        msg: dict[str, Any],
        timestamp_ms: int,
        part: NormalizedPart,
    ) -> int:
        state = self._state
        folded = try_append_to_assistant_group(
            state.messages,
            current_assistant_index=state.current_assistant_index,
            parts=(part,),
            blocking_part_types=("tool",),
            on_message=lambda existing: self._apply_assistant_metadata(existing, msg),
        )
        if folded is not None:
            return folded

        message = build_message(
            message_id=message_id,
            role="assistant",
            agent="claude",
            time_created=timestamp_ms,
            parts=[part],
        )
        self._apply_assistant_metadata(message, msg)
        state.messages.append(message)
        return len(state.messages) - 1

    def _attach_tool_call(
        self,
        *,
        message_id: str,
        msg: dict[str, Any],
        timestamp_ms: int,
        tool_part: NormalizedPart,
    ) -> tuple[int, int]:
        state = self._state
        if state.latest_assistant_text_index is not None:
            message = state.messages[state.latest_assistant_text_index]
            message["parts"].append(tool_part)
            self._apply_assistant_metadata(message, msg)
            return state.latest_assistant_text_index, len(message["parts"]) - 1

        message = build_message(
            message_id=message_id,
            role="assistant",
            agent="claude",
            time_created=timestamp_ms,
            mode="tool",
            parts=[tool_part],
        )
        self._apply_assistant_metadata(message, msg)
        state.messages.append(message)
        return len(state.messages) - 1, 0

    def _extract_tool_state_updates(self, tool_use_result: Any) -> dict[str, Any]:
        if not isinstance(tool_use_result, dict):
            return {}

        updates: dict[str, Any] = {}
        success = tool_use_result.get("success")
        if isinstance(success, bool):
            updates["status"] = "success" if success else "error"
        command_name = tool_use_result.get("commandName")
        if command_name:
            updates["meta"] = {"commandName": command_name}
        return updates

    def _backfill_tool_output(
        self,
        *,
        call_id: str,
        output_parts: list[NormalizedPart],
        state_updates: dict[str, Any] | None = None,
    ) -> bool:
        state = self._state
        tool_part = backfill_tool_state(
            state.messages,
            state.pending_tool_calls,
            call_id=call_id,
            output_parts=output_parts,
            state_updates=state_updates,
        )
        if tool_part is None:
            return False
        tool_state = tool_part.setdefault("state", {})
        if output_parts and "status" not in tool_state:
            tool_state["status"] = "completed"
        return bool(output_parts or state_updates)

    def _resolve_tool_call_id(self, data: dict[str, Any], item: dict[str, Any]) -> str:
        tool_call_id = str(item.get("tool_use_id", "")).strip()
        if tool_call_id:
            return tool_call_id
        source_uuid = str(data.get("sourceToolAssistantUUID", "")).strip()
        if not source_uuid:
            return ""
        tool_call_ids = self._state.assistant_uuid_to_tool_calls.get(source_uuid, [])
        return tool_call_ids[0] if len(tool_call_ids) == 1 else ""

    def _append_assistant_record(self, data: dict[str, Any]) -> None:
        state = self._state
        msg = data.get("message", {})
        timestamp_ms = self._parse_timestamp_ms(data)
        raw_content = msg.get("content", [])
        tool_call_ids: list[str] = []

        if isinstance(raw_content, list):
            for item in raw_content:
                if not isinstance(item, dict):
                    continue
                part_type = item.get("type")
                message_id = str(data.get("uuid", ""))
                if part_type == "thinking":
                    text = str(item.get("thinking", ""))
                    if text.strip():
                        state.current_assistant_index = self._append_assistant_reasoning(
                            message_id=message_id,
                            msg=msg,
                            timestamp_ms=timestamp_ms,
                            part=build_text_part(text, timestamp_ms, part_type="reasoning"),
                        )
                    continue
                if part_type == "text":
                    text = str(item.get("text", ""))
                    if text.strip():
                        state.current_assistant_index = self._append_assistant_text(
                            message_id=message_id,
                            msg=msg,
                            timestamp_ms=timestamp_ms,
                            part=build_text_part(text, timestamp_ms),
                        )
                        state.latest_assistant_text_index = state.current_assistant_index
                    continue
                if part_type != "tool_use":
                    continue

                tool_name = str(item.get("name", "")).strip()
                tool_call_id = str(item.get("id", "")).strip()
                if tool_name == "TodoWrite" and tool_call_id:
                    state.ignored_tool_call_ids.add(tool_call_id)
                    continue

                message_index, part_index = self._attach_tool_call(
                    message_id=message_id,
                    msg=msg,
                    timestamp_ms=timestamp_ms,
                    tool_part=self._build_tool_part(item, timestamp_ms),
                )
                state.current_assistant_index = message_index
                if tool_call_id:
                    state.pending_tool_calls[tool_call_id] = (message_index, part_index)
                    tool_call_ids.append(tool_call_id)

        if tool_call_ids:
            state.assistant_uuid_to_tool_calls[str(data.get("uuid", ""))] = tool_call_ids

    def _append_user_record(self, data: dict[str, Any]) -> None:
        state = self._state
        msg = data.get("message", {})
        timestamp_ms = self._parse_timestamp_ms(data)
        content = msg.get("content", "")

        if isinstance(content, str):
            parts = self._normalize_user_text_parts(content, timestamp_ms)
            if not parts:
                return
            state.messages.append(
                build_message(
                    message_id=str(data.get("uuid", "")),
                    role="user",
                    time_created=timestamp_ms,
                    parts=parts,
                )
            )
            state.reset_assistant_group()
            return
        if not isinstance(content, list):
            state.reset_assistant_group()
            return

        visible_parts = self._normalize_user_text_parts(content, timestamp_ms)
        tool_state_updates = self._extract_tool_state_updates(data.get("toolUseResult"))
        for item in content:
            if not isinstance(item, dict) or item.get("type") != "tool_result":
                continue
            tool_call_id = self._resolve_tool_call_id(data, item)
            if tool_call_id and tool_call_id in state.ignored_tool_call_ids:
                continue

            output_parts = self._normalize_tool_output(item.get("content"), timestamp_ms)
            if self._backfill_tool_output(
                call_id=tool_call_id,
                output_parts=output_parts,
                state_updates=tool_state_updates,
            ):
                continue
            fallback_message = build_fallback_tool_message(
                message_id=str(data.get("uuid", "")),
                output_parts=output_parts,
                time_created=timestamp_ms,
                tool_call_id=tool_call_id or None,
            )
            if fallback_message:
                state.messages.append(fallback_message)

        if visible_parts:
            state.messages.append(
                build_message(
                    message_id=str(data.get("uuid", "")),
                    role="user",
                    time_created=timestamp_ms,
                    parts=visible_parts,
                )
            )
        state.reset_assistant_group()

File: rust/src/claude_transcript.rs
use crate::message_assembly::{backfill, fold_assistant};
use crate::session::{
    Message, Part, Session, SessionData, Stats, TextPart, ToolPart, parse_timestamp,
};
use crate::value::{field, string, text, truthy};
use serde_json::{Map, Value, json};
use std::collections::{HashMap, HashSet};

#[derive(Default)]
struct Decoder {
    messages: Vec<Message>,
    pending: HashMap<String, (usize, usize)>,
    ignored: HashSet<String>,
    assistant_calls: HashMap<String, Vec<String>>,
    current: Option<usize>,
    latest_text: Option<usize>,
}

fn output_parts(content: &Value, timestamp: i64) -> Vec<Part> {
    let texts = match content {
        Value::Null => vec![],
        Value::Array(items) => items
            .iter()
            .filter_map(|item| match item {
                Value::Object(_) => Some(
                    item.get("text")
                        .or_else(|| item.get("content"))
                        .map(string)
                        .unwrap_or_default(),
                ),
                Value::String(value) => Some(value.clone()),
                _ => None,
            })
            .collect(),
        _ => vec![string(content)],
    };
    texts
        .into_iter()
        .filter(|text| !text.trim().is_empty())
        .map(|text| Part::text(text, timestamp))
        .collect()
}

fn user_parts(content: &Value, timestamp: i64) -> Vec<Part> {
    match content {
        Value::String(_) => output_parts(content, timestamp),
        Value::Array(items) => items
            .iter()
            .filter_map(|item| {
                let text = match item {
                    Value::Object(_) if item["type"] != "tool_result" => field(item, "text"),
                    Value::String(value) => value.clone(),
                    _ => return None,
                };
                (!text.trim().is_empty()).then(|| Part::text(text, timestamp))
            })
            .collect(),
        _ => vec![],
    }
}

fn apply_metadata(message: &mut Message, source: &Value) {
    if truthy(&source["model"]) && message.model.as_ref().is_none_or(|value| !truthy(value)) {
        message.model = Some(source["model"].clone());
    }
    if message.tokens.is_empty()
        && let Some(usage) = source["usage"].as_object()
    {
        message.tokens = usage.clone();
    }
}

impl Decoder {
    fn reset(&mut self) {
        self.current = None;
        self.latest_text = None;
    }

    fn assistant_part(
        &mut self,
        id: &str,
        source: &Value,
        timestamp: i64,
        part: Part,
        reasoning: bool,
    ) {
        let mut parts = vec![part];
        let index = fold_assistant(&mut self.messages, self.current, &mut parts, reasoning)
            .unwrap_or_else(|| {
                let mut message = Message::new(id.to_owned(), "assistant", timestamp, parts);
                message.agent = Some("claude".into());
                self.messages.push(message);
                self.messages.len() - 1
            });
        apply_metadata(&mut self.messages[index], source);
        self.current = Some(index);
        if !reasoning {
            self.latest_text = Some(index);
        }
    }

    fn assistant(&mut self, record: &Value, source: &Value, timestamp: i64) {
        let Some(items) = source["content"].as_array() else {
            return;
        };
        let id = field(record, "uuid");
        let mut calls = Vec::new();
        for item in items {
            match text(&item["type"]) {
                "text" | "thinking" => {
                    let reasoning = item["type"] == "thinking";
                    let body = field(item, if reasoning { "thinking" } else { "text" });
                    if body.trim().is_empty() {
                        continue;
                    }
                    let part = if reasoning {
                        Part::Reasoning(TextPart {
                            text: body,
                            time_created: timestamp,
                        })
                    } else {
                        Part::text(body, timestamp)
                    };
                    self.assistant_part(&id, source, timestamp, part, reasoning);
                }
                "tool_use" => {
                    let name = field(item, "name");
                    let raw_id = field(item, "id");
                    let call_id = raw_id.trim().to_owned();
                    if name.trim() == "TodoWrite" && !call_id.is_empty() {
                        self.ignored.insert(call_id);
                        continue;
                    }
                    let input = item.get("input").cloned().unwrap_or_else(|| json!({}));
                    let part = Part::Tool(Box::new(ToolPart {
                        subagent_type: None,
                        tool: name.clone(),
                        call_id: raw_id,
                        title: format!("Tool: {name}"),
                        state: json!({"input": input, "output": null})
                            .as_object()
                            .unwrap()
                            .clone(),
                        time_created: timestamp,
                        nickname: None,
                        subagent_id: None,
                    }));
                    let index = if let Some(index) = self.latest_text {
                        self.messages[index].parts.push(part);
                        index
                    } else {
                        let mut message =
                            Message::new(id.clone(), "assistant", timestamp, vec![part]);
                        message.agent = Some("claude".into());
                        message.mode = Some("tool".into());
                        self.messages.push(message);
                        self.messages.len() - 1
                    };
                    apply_metadata(&mut self.messages[index], source);
                    self.current = Some(index);
                    if !call_id.is_empty() {
                        self.pending.insert(
                            call_id.clone(),
                            (index, self.messages[index].parts.len() - 1),
                        );
                        calls.push(call_id);
                    }
                }
                _ => (),
            }
        }
        if !calls.is_empty() {
            self.assistant_calls.insert(id, calls);
        }
    }

    fn fallback(&mut self, id: String, timestamp: i64, parts: Vec<Part>, call_id: Option<String>) {
        if !parts.is_empty() {
            let mut message = Message::new(id, "tool", timestamp, parts);
            message.tool_call_id = call_id;
            self.messages.push(message);
        }
    }

    fn user(&mut self, record: &Value, source: &Value, timestamp: i64) -> crate::Result<()> {
        let content = &source["content"];
        let visible = user_parts(content, timestamp);
        if content.is_string() && visible.is_empty() {
            return Ok(());
        }
        if let Some(items) = content.as_array() {
            let mut updates = Map::new();
            let result = &record["toolUseResult"];
            if let Some(success) = result["success"].as_bool() {
                updates.insert(
                    "status".into(),
                    if success { "success" } else { "error" }.into(),
                );
            }
            if truthy(&result["commandName"]) {
                updates.insert("meta".into(), json!({"commandName": result["commandName"]}));
            }
            for item in items.iter().filter(|item| item["type"] == "tool_result") {
                let mut call_id = field(item, "tool_use_id").trim().to_owned();
                if call_id.is_empty() {
                    let uuid = field(record, "sourceToolAssistantUUID").trim().to_owned();
                    if let Some(calls) = self
                        .assistant_calls
                        .get(&uuid)
                        .filter(|calls| calls.len() == 1)
                    {
                        call_id = calls[0].clone();
                    }
                }
                if self.ignored.contains(&call_id) {
                    continue;
                }
                let parts = output_parts(&item["content"], timestamp);
                if let Some(tool) = backfill(
                    &mut self.messages,
                    &self.pending,
                    &call_id,
                    &parts,
                    Some(updates.clone()),
                )? {
                    if !parts.is_empty() {
                        tool.state.entry("status").or_insert("completed".into());
                    }
                } else {
                    self.fallback(
                        field(record, "uuid"),
                        timestamp,
                        parts,
                        (!call_id.is_empty()).then_some(call_id),
                    );
                }
            }
        }
        if !visible.is_empty() {
            self.messages.push(Message::new(
                field(record, "uuid"),
                "user",
                timestamp,
                visible,
            ));
        }
        self.reset();
        Ok(())
    }

    fn record(&mut self, record: &Value) -> crate::Result<()> {
        if record["isMeta"] == true {
            return Ok(());
        }
        let source = &record["message"];
        if !source.is_object() {
            return Ok(());
        }
        let timestamp =
            parse_timestamp(text(&record["timestamp"])).map_or(0, |time| time.as_millisecond());
        match text(&record["type"]) {
            "assistant" => self.assistant(record, source, timestamp),
            "user" => self.user(record, source, timestamp)?,
            "tool_result" => {
                self.fallback(
                    field(record, "uuid"),
                    timestamp,
                    output_parts(&source["content"], timestamp),
                    None,
                );
                self.reset();
            }
            _ => (),
        }
        Ok(())
    }
}

pub fn read(session: &Session) -> crate::Result<SessionData> {
    let mut decoder = Decoder::default();
    crate::jsonl::scan(&session.source_path, |record| decoder.record(&record))?;
    let mut stats = Stats {
        message_count: decoder.messages.len(),
        ..Stats::default()
    };
    for message in &decoder.messages {
        stats.add_tokens(
            message.tokens.get("input_tokens").unwrap_or(&Value::Null),
            message.tokens.get("output_tokens").unwrap_or(&Value::Null),
        )?;
    }
    Ok(session.payload(decoder.messages, stats))
}
